### PR TITLE
Remove no-default-scrape-port featureFlag

### LINKED ISSUE #13959: --enable-feature: Consider removing no-default-scrape-port
### Proposal

Please consider if the behavior enabled with `--enable-feature=no-default-scrape-port` should become the default behavior in Prometheus 3.x /cc @roidelapluie @LeviHarrison

### PR DESCRIPTION
**Pull request by alex-kattathra-johnson**
Fixes #13959

### CODE REVIEW

**Comment by jan--f** [comment]
This has to wait till the prometheus 3 dev branch is available.

**Review by jan--f — COMMENTED** [review]
I'm not sure adding this feature as a default is the best solution here. If someone tries to pass a feature flag, the default will be overwritten.

Afaiu the issue, the goal is for the feature flag to be enabled by default, i.e. there should be a mechanism to get the old behavior back.
If that is correct we need a way to get the default scrape port behavior back,
If I'm wrong and we don't want an option for the old behavior, this can be implemented similar to other feature promotions to stable, like https://github.com/prometheus/prometheus/pull/10121
This removes the unexpected behavior when user pass a feature flag.
Wdyt @beorn7 @roidelapluie ?

**Comment by beorn7** [comment]
I think we want to remove the feature flag entirely and have no way to bring the old behavior back.
But @roidelapluie has more context and should confirm.

**Comment by jan--f** [comment]
@alex-kattathra-johnson Seems like you are also trying to merge the `main` branch here. Can you base your changes on the `release-3.0` branch please.

**Comment by bboreham** [comment]
I fixed up the commits, but it is still doing the less-desired thing.

**Comment by jan--f** [comment]
We have merge `release-3.0` into `main`. The base branch for this PR should be changed to `main` now.

**Comment by alex-kattathra-johnson** [comment]
@bboreham : I might've removed your fixed up commit from earlier when trying to rebase on main.

**Review by jan--f — CHANGES_REQUESTED** [review]
The goal here is to enable the behavior by default. Currently this passes a default value to the feature flag list. While this achieves the default on behavior, its would be better to implement the on by default behavior similar to https://github.com/prometheus/prometheus/pull/10121/files.
The feature flag should be removed completely with no way to get the old behavior back.

**Comment by alex-kattathra-johnson** [comment]
@jan--f : After remove the feature flag, which features should we have enabled? Should we enable all the features that could potentially be set via the coma separated feature flag?

**Review by jan--f — CHANGES_REQUESTED** [review]
I might have made it unclear, but the intention is not to remove the feature flag option. The goal is to remove a single feature flag and enable it by default. The feature flag is called `no-default-scrape-port`.
You can follow the same pattern as in #10121.

File: scrape/target.go
Comment (jan--f) [inline scrape/target.go:461]: Iiuc this is comment line is no longer accurate?

File: scrape/target.go
Comment (jan--f) [inline scrape/target.go:476]: A comment here might still be good? E.g.
`If it's an address with a trailing port that corresponds to the scheme, remove the the port.`

File: cmd/promtool/main.go
Comment (jan--f) [inline cmd/promtool/main.go:294]: Question for a maintainer, maybe @beorn7 : Should this flag be dropped or should it be kept, but with no feature to enable for now?

**Review by jan--f — COMMENTED** [review]
Overall this looks good I think. Only a few more questions.

File: cmd/promtool/main.go
Comment (beorn7) [reply cmd/promtool/main.go:294]: I would remove it entirely. It's easy to bring back, but having this command line flag without any feature flags sounds weird. The other side of the argument would be that it is almost certain that we'll have feature flags in promtool, too, before we release v4. @dgl should make the call here.

File: scrape/target.go
Comment (beorn7) [reply scrape/target.go:476]: The weird thing is that I couldn't see this behavior in practice. Let me investigate more.

File: scrape/target.go
Comment (beorn7) [reply scrape/target.go:461]: Yes, the comment needs an update.

Comment (beorn7) [reply scrape/target.go:461]: But maybe we don't need the whole logic anymore. Let me investigate.

**Comment by alex-kattathra-johnson** [comment]
It seems someone else has created a PR by copying the changes from here, but I made the recommended changes and rebased on `main` anyhow. Do with it as you please! :)

File: cmd/promtool/main.go
Comment (beorn7) [reply cmd/promtool/main.go:294]: Yet another argument is to not needlessly break users that still pass in the feature flag. It shouldn't be a problem to activate a feature flag that is now default.

This might convince me to keep the `--enable-feature` flag in `promtool`, just without any possible option that would do anything (as it was done in #14958, the PR that weirdly was created mirroring this PR).

Comment (alex-kattathra-johnson) [reply cmd/promtool/main.go:294]: That's a good reason, IMO. I'll add a dummy flag back in.

File: scrape/target.go
Comment (beorn7) [reply scrape/target.go:476]: Got it now: So what happens is that we remove the default port from the address label, but _not_ from the instance label. I am actually not sure that this was intentional when the feature flag was implemented. It's more an artifact of modifying `addr` or (like in this part of the code) just setting the `model.AddressLabel` directly.

My current gut feeling is that we should _not_ remove the port at all, even if it is the default, or in other words, we really leave it to the user if the port should be there or not.

**Comment by beorn7** [comment]
Sorry for confusion. An almost similar PR was created recently, see #14958, and some discussion happened over there. The discussion over there is relevant, but the changes should happen in this PR, which is the original one.

**Comment by jan--f** [comment]
> Sorry for confusion. An almost similar PR was created recently, see #14958, and some discussion happened over there. The discussion over there is relevant, but the changes should happen in this PR, which is the original one.

As mentioned on the other PR, this was my mistake. After this PR saw no activity for 2 month I gave the thumbs up when someone else wanted to take a stab at it.

File: cmd/promtool/main.go
Comment (beorn7) [inline cmd/promtool/main.go:331]: Nit for brevity:

```suggestion
				fmt.Printf("  WARNING: --enable-feature is currently a no-op")
```

**Review by beorn7 — CHANGES_REQUESTED** [review]
Few remaining nits about word smithing.

File: cmd/promtool/main.go
Comment (alex-kattathra-johnson) [reply cmd/promtool/main.go:294]: loled at "we don't have to restrict nothingness even further"

**Review by beorn7 — APPROVED** [review]
Thank you very much.

Before merging, I would like to hear from @bboreham and @roidelapluie , though.

**Comment by beorn7** [comment]
And of course @jan--f should also agree that this is now what it should be.

**Comment by bboreham** [comment]
I took a look round the targets on a Prombench instance of Prometheus.
All of them appeared to have a port supplied via service discovery or via config.
My understanding of this PR is that nothing will change with those targets.
Therefore I am ok going ahead.

**Comment by beorn7** [comment]
Yes, at least that's the intention of this PR: If a port is explicitly provided, Prometheus won't interfere with that in any way.

**Comment by beorn7** [comment]
I'll merge this now to get the next beta release rolling. We can still rollback before we do an RC.

**Comment by flav-code** [comment]
Hello,
I think this broke my stats...
image

**Comment by beorn7** [comment]
You can adjust your scrape config so that the instance label has the desired shape. See also the [migration guide](https://prometheus.io/docs/prometheus/latest/migration/#flags).

**Comment by beorn7** [comment]
There are many ways to address this. But a merged PR on GH is not the right place for usage support. Please refer to the [community channels](https://prometheus.io/community/) for better ways of getting community support.

**Pull request merged**
